File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/actors/mobs/Tengu.java
/*
 * Pixel Dungeon
 * Copyright (C) 2012-2014  Oleg Dolya
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */
package com.watabou.pixeldungeon.actors.mobs;

import com.nyrds.pixeldungeon.items.common.ItemFactory;
import com.nyrds.pixeldungeon.levels.objects.LevelObjectsFactory;
import com.nyrds.pixeldungeon.levels.objects.Trap;
import com.nyrds.pixeldungeon.mechanics.NamedEntityKind;
import com.nyrds.pixeldungeon.ml.R;
import com.nyrds.pixeldungeon.mobs.common.IZapper;
import com.nyrds.platform.audio.Sample;
import com.nyrds.platform.util.StringsManager;
import com.watabou.pixeldungeon.Assets;
import com.watabou.pixeldungeon.Badges;
import com.watabou.pixeldungeon.Badges.Badge;
import com.watabou.pixeldungeon.Dungeon;
import com.watabou.pixeldungeon.actors.Actor;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.CharUtils;
import com.watabou.pixeldungeon.actors.blobs.ToxicGas;
import com.watabou.pixeldungeon.actors.buffs.Poison;
import com.watabou.pixeldungeon.actors.hero.Hero;
import com.watabou.pixeldungeon.actors.hero.HeroClass;
import com.watabou.pixeldungeon.effects.CellEmitter;
import com.watabou.pixeldungeon.effects.Speck;
import com.watabou.pixeldungeon.items.TomeOfMastery;
import com.watabou.pixeldungeon.items.keys.SkeletonKey;
import com.watabou.pixeldungeon.items.potions.PotionOfHealing;
import com.watabou.pixeldungeon.items.scrolls.ScrollOfMagicMapping;
import com.watabou.pixeldungeon.levels.Level;
import com.watabou.pixeldungeon.mechanics.Ballistica;
import com.watabou.pixeldungeon.utils.Utils;
import com.watabou.utils.Random;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

public class Tengu extends Boss implements IZapper {

	private static final int JUMP_DELAY = 5;
	
	public Tengu() {
		
		hp(ht(120));
		exp = 20;
		baseDefenseSkill = 20;
		baseAttackSkill  = 20;
		dmgMin = 8;
		dmgMax = 15;
		dr = 5;

		addResistance( ToxicGas.class );
		addResistance( Poison.class );

		final HeroClass heroClass = Dungeon.heroClass;

		if ( heroClass != HeroClass.NECROMANCER && heroClass != HeroClass.GNOLL){
			collect(new TomeOfMastery());
		}

		if( heroClass == HeroClass.GNOLL) {
			collect(ItemFactory.itemByName("TenguLiver"));
		}

		collect(new SkeletonKey());
	}
	
	private int timeToJump = JUMP_DELAY;

	@Override
	public void die(@NotNull NamedEntityKind cause) {
		super.die(cause);
		
		Badges.validateBossSlain(Badge.BOSS_SLAIN_2);

        say(StringsManager.getVar(R.string.Tengu_Info1));
	}
	
	@Override
	public boolean getCloser(int target) {
		if (level().fieldOfView[target]) {
			jump();
			return true;
		} else {
			return super.getCloser( target );
		}
	}
	
	@Override
    public boolean canAttack(@NotNull Char enemy) {
		return Ballistica.cast( getPos(), enemy.getPos(), false, true ) == enemy.getPos();
	}
	
	@Override
	public void doAttack(Char enemy) {
		timeToJump--;
		if (timeToJump <= 0 && adjacent(enemy)) {
			jump();
		} else {
			super.doAttack( enemy );
		}
	}
	
	private void jump() {
		timeToJump = JUMP_DELAY;

		final Level level1 = level();

		for (int i = 0; i < 4; i++) {
			int trapPos = level1.getRandomTerrain((level, cell) -> level.getTopLevelObject(cell) instanceof Trap);

			if (level1.cellValid(trapPos)) {
				Trap tr = (Trap) level1.getTopLevelObject(trapPos);
				tr.reactivate(LevelObjectsFactory.POISON_TRAP, 1);
				ScrollOfMagicMapping.discover( trapPos );
			} else {
				break;
			}
		}

		ArrayList<Integer> candidates = new ArrayList<>();
		int enemyPos = getEnemy().getPos();
		for(int i = 0; i< level1.getLength(); ++i) {
			if(level1.fieldOfView[i]
					&& level1.passable[i]
					&& !level1.adjacent( i, enemyPos )
					&& Actor.findChar(i) == null) {
				candidates.add(i);
			}
		}

		if(candidates.isEmpty()) {
			PotionOfHealing.heal(this, 0.1f);
			spend( 1 / speed() );
			return;
		}

		int newPos = candidates.get(Random.index(candidates));

		getSprite().move( getPos(), newPos );
		move( newPos );
		
		if (CharUtils.isVisible(this)) {
			CellEmitter.get( newPos ).burst( Speck.factory( Speck.WOOL ), 6 );
			Sample.INSTANCE.play( Assets.SND_PUFF );
		}
		
		spend( 1 / speed() );
	}
	
	@Override
	public void notice() {
		super.notice();
        String tenguYell = StringsManager.getVar(R.string.Tengu_Info2);
		final Hero hero = Dungeon.hero;
		final HeroClass heroClass = hero.getHeroClass();

		if (heroClass.getGender() == Utils.FEMININE) {
            tenguYell = StringsManager.getVar(R.string.Tengu_Info3);
		}
		yell(Utils.format(tenguYell, heroClass.title()));
	}	
}

File: scripts/mobs/Tengu.lua
--[[
  Tengu (batch 17d-1, was actors/mobs/Tengu.java).
  Prison boss: while the enemy is visible Tengu jumps instead of walking
  (getCloser hook), and every 5th attack next to the hero is a jump too
  (dodge, doAttack hook). Jumps re-arm up to 4 random traps into poison
  traps and reveal them; with no free cell to land on Tengu falls back to
  a potion-style 10% heal.
]]
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"
local JUMP_DELAY = 5

-- java Level.adjacent is a raw cell-diff check (rows can "wrap"); port it
-- verbatim instead of a proper chebyshev
local function adjacent(level, a, b)
    local diff = math.abs(a - b)
    local W = level:getWidth()
    return diff == 1 or diff == W or diff == W + 1 or diff == W - 1
end

local function rearmTraps(level)
    local traps = {}
    for _, obj in pairs(level:getLevelObjects()) do
        if obj:isTrap() then
            traps[#traps + 1] = obj
        end
    end

    for _ = 1, 4 do
        if #traps == 0 then
            break
        end
        local trap = traps[math.random(#traps)]
        trap:reactivate("PoisonTrap", 1)
        RPD.ScrollOfMagicMapping:discover(trap:getPos())
    end
end

local function jump(self)
    local data = mob.restoreData(self)
    data.timeToJump = JUMP_DELAY

    local level = RPD.Dungeon.level
    rearmTraps(level)

    local enemyPos = self:getEnemy():getPos()
    local candidates = {}
    for i = 0, level:getLength() - 1 do
        if level.fieldOfView[i + 1]
                and level.passable[i + 1]
                and not adjacent(level, i, enemyPos)
                and RPD.Actor:findChar(i) == nil then
            candidates[#candidates + 1] = i
        end
    end

    if #candidates == 0 then
        -- java fell back to PotionOfHealing.heal(this, 0.1f)
        self:heal(math.floor(self:ht() * 0.1), self)
        self:detachBuff("Poison")
        self:detachBuff("Cripple")
        self:detachBuff("Weakness")
        self:detachBuff("Bleeding")
        self:spend(1 / self:speed())
        return
    end

    local newPos = candidates[math.random(#candidates)]
    self:move(newPos)

    if RPD.CharUtils:isVisible(self) then
        RPD.Sfx.CellEmitter:get(newPos):burst(RPD.Sfx.Speck:factory(RPD.Sfx.Speck.WOOL), 6)
        RPD.playSound("snd_puff")
    end

    self:spend(1 / self:speed())
end

return mob.init{
    spawn = function(self, level)
        local heroClass = RPD.Dungeon.hero:getHeroClass():name()
        local bag = self:getBelongings()

        if heroClass ~= "NECROMANCER" and heroClass ~= "GNOLL" and heroClass ~= "DOCTOR"
                and not bag:getItem("TomeOfMastery") then
            self:collect(RPD.item("TomeOfMastery"))
        end

        if heroClass == "GNOLL" and not bag:getItem("TenguLiver") then
            self:collect(RPD.item("TenguLiver"))
        end
    end,

    getCloser = function(self, target, ignorePets)
        local level = RPD.Dungeon.level
        if level.fieldOfView[target + 1] then
            -- refund doStepTo's pre-charge; jump() spends the real cost
            self:spend(-1 / self:speed())
            jump(self)
            return true
        end
        return nil
    end,

    doAttack = function(self, enemy)
        local data = mob.restoreData(self)
        data.timeToJump = (data.timeToJump or JUMP_DELAY) - 1
        if data.timeToJump <= 0
                and adjacent(RPD.Dungeon.level, self:getPos(), enemy:getPos()) then
            jump(self)
            return true
        end
        return nil
    end,

    notice = function(self)
        local heroClass = RPD.Dungeon.hero:getHeroClass()
        local id = "Tengu_Info2"
        if heroClass:getGender() == 2 then
            id = "Tengu_Info3"
        end
        self:yell(RPD.textById(id):format(heroClass:title()))
    end,

    die = function(self, cause)
        RPD.CharUtils:validateBossSlain("BOSS_SLAIN_2")
        self:say(RPD.textById("Tengu_Info1"))
    end,
}
